### ISSUE TITLE
ivy port

### ISSUE THREAD

**Issue by justbur** [tmalsburg/helm-bibtex#101]
Hi, I started working on a port of your package to use [ivy](https://github.com/abo-abo/swiper) as a backend for the completion. I was wondering if you would want to incorporate the work here, which would involve refactoring some of the code to make it less dependent on helm, or if I should just put it in a new repo? It's not that hard actually. In a little bit of time, I got something working already. You can see my fork for the basic stuff I did.

**Comment by tmalsburg** [comment]
There was recently a discussion on reddit about making completion sources independent from the user interface (helm, ivy, ...).  The goal would be to make efforts like this one easier, but unfortunately there is currently no standard interface for such generic sources.

I'm not familiar with ivy and therefore don't know how difficult it would be to factor out the non-helm-related logic for use with ivy.  If it can be done without sacrificing maintainability of the code, I'm totally in favor of doing it.  Of course it's possible that the code would even become cleaner.  Happy to talk details.

**Comment by justbur** [comment]
If you're familiar with completing-read, ivy-read has a very similar
interface with some added features, most notably the ability to specify one
or more actions to act on selected candidates.

From what I saw looking today, the main changes I would have to make are
related to the formatting of candidates and the action functions. The
"transformer" functions have different interfaces between helm and ivy. The
action functions seem to as well. In ivy, actions receive either the cdr of
the candidate (if there is one) or just a string representing the candidate
itself. It appears that your action functions work differently.

The main work is in generating the list of candidates, which I'd like to
use. The rest seems relatively straightforward. Maybe you (or we) could
factor out the backend parsing stuff into a generic "bibtex-completion"
library that helm-bibtex and ivy-bibtex could depend on?

**Comment by tmalsburg** [comment]
Do you think it would make sense to share any formatting code?  Or should the backend only be concerned with parsing bibtex and finding PDFs?

Regarding the actions, it sounds as if we could define them in the backend and just add a thin facade in helm-bibtex and ivy-bibtex.

**Comment by justbur** [comment]
Check out what I did with [my fork](https://github.com/justbur/helm-bibtex).

I copied helm-bibtex.el to bibtex-completion.el and made most of the actions and commands more generic. The only thing I haven't done yet is format the ivy candidates. The "formatter" function depends on helm and works in a way that is not easily portable to ivy. I will either write a new one or pull out the generic bit.

See ivy-bibtex.el for how to use the library now. It's pretty simple. A similar thing could be done on the helm-bibtex.el side, but I haven't done it yet.
